src/coap_tinydtls.c

```diff
@@ -304,34 +328,142 @@ get_psk_info(struct dtls_context_t *dtls_context,
   return dtls_alert_fatal_create(fatal_error);
 }
 
-static dtls_handler_t cb = {
+#ifdef DTLS_ECC
+static int
+get_ecdsa_key(struct dtls_context_t *dtls_context,
+              const session_t *dtls_session UNUSED,
+              const dtls_ecdsa_key_t **result) {
+  static dtls_ecdsa_key_t ecdsa_key;
+  coap_tiny_context_t *t_context =
+                  (coap_tiny_context_t *)dtls_get_app_data(dtls_context);
+
+  ecdsa_key.curve = DTLS_ECDH_CURVE_SECP256R1;
+  ecdsa_key.priv_key = t_context->priv_key->s;
+  ecdsa_key.pub_key_x = t_context->pub_key->s;
+  ecdsa_key.pub_key_y = &t_context->pub_key->s[DTLS_EC_KEY_SIZE];
+
+  *result = &ecdsa_key;
+  return 0;
+}
+
+/* first part of Raw public key, the is the start of the Subject Public Key */
+static const unsigned char cert_asn1_header[] = {
+  0x30, 0x59, /* SEQUENCE, length 89 bytes */
+    0x30, 0x13, /* SEQUENCE, length 19 bytes */
+      0x06, 0x07, /* OBJECT IDENTIFIER ecPublicKey (1 2 840 10045 2 1) */
+        0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x02, 0x01,
+      0x06, 0x08, /* OBJECT IDENTIFIER prime256v1 (1 2 840 10045 3 1 7) */
+        0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x03, 0x01, 0x07,
+      0x03, 0x42, 0x00, /* BIT STRING, length 66 bytes, 0 bits unused */
+         0x04 /* uncompressed, followed by the r and s values of the public key */
+};
+#define DTLS_CE_LENGTH (3 + 27 + key_size + key_size)
+#define DTLS_EC_SUBJECTPUBLICKEY_SIZE (2 * key_size + sizeof(cert_asn1_header))
+
+static int
+verify_ecdsa_key(struct dtls_context_t *dtls_context UNUSED,
+                 const session_t *dtls_session UNUSED,
+                 const uint8_t *other_pub_x,
+                 const uint8_t *other_pub_y,
+                 size_t key_size) {
+  coap_tiny_context_t *t_context =
+                  (coap_tiny_context_t *)dtls_get_app_data(dtls_context);
+  if (t_context && t_context->setup_data.validate_cn_call_back) {
+    /* Need to build asn.1 certificate - code taken from tinydtls */
+    uint8 *p;
+    uint8 buf[DTLS_CE_LENGTH];
+    coap_session_t *c_session;
+    coap_address_t remote_addr;
+
+    /* Certificate
+     *
+     * Start message construction at beginning of buffer. */
+    p = buf;
+
+    /* length of this certificate */
+    dtls_int_to_uint24(p, DTLS_EC_SUBJECTPUBLICKEY_SIZE);
+    p += sizeof(uint24);
+
+    memcpy(p, &cert_asn1_header, sizeof(cert_asn1_header));
+    p += sizeof(cert_asn1_header);
+
+    memcpy(p, other_pub_x, key_size);
+    p += key_size;
+
+    memcpy(p, other_pub_y, key_size);
+    p += key_size;
+
+    assert(p <= (buf + sizeof(buf)));
+
+    get_session_addr(dtls_session, &remote_addr);
+    c_session = coap_session_get_by_peer(t_context->c_context,
+                                         &remote_addr, dtls_session->ifindex);
+    if (!c_session)
+      return -3;
+    if (!t_context->setup_data.validate_cn_call_back("RPK", buf, p-buf,
```

Comment:
As the string "RPK" is used for gnutls and tinydtls (and possibly other libraries in case there will be RPK support), we might want to define a constant that is used in both files. As only one of coap_{tlslibrary}.c is used, a pre-processor symbol in `coap_internal.h` would do, for example.

Reply:
Defined COAP_DTLS_RPK_CERT_CN but made it app visible for handling cn_validation checks.